src/seat/keyboard/repeat.rs

```diff
@@ -0,0 +1,242 @@
+        let disabled = &mut self.disabled;
+
+        let mut reregister = false;
+
+        // Check if the key repeat should stop
+        let channel_pa = self
+            .channel
+            .process_events(readiness, token, |event, _| {
+                match event {
+                    channel::Event::Msg(message) => {
+                        match message {
+                            RepeatMessage::StopRepeat => {
+                                key.take();
+                            }
+
+                            RepeatMessage::StartRepeat(mut event) => {
+                                // Update time for next event
+                                event.time += *delay_mut as u32;
+                                key.replace(event);
+                                reregister = true;
+
+                                // Schedule a new press event in the timer.
+                                timer.set_duration(Duration::from_millis(*delay_mut));
+                            }
+
+                            RepeatMessage::RepeatInfo(info) => {
+                                match info {
+                                    // Store the repeat time, using it for the next repeat sequence.
+                                    RepeatInfo::Repeat { rate, delay } => {
+                                        // Number of repetitions per second / 1000 ms
+                                        *gap = (rate.get() / 1000) as u64;
+                                        *delay_mut = delay as u64;
+                                        *disabled = false;
+                                        timer.set_duration(Duration::from_millis(*delay_mut));
+                                    }
+
+                                    RepeatInfo::Disable => {
+                                        // Compositor will send repeat events manually, cancel all repeating events
+                                        key.take();
+                                        *disabled = true;
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    channel::Event::Closed => {
+                        removed = true;
+                    }
+                }
+            })
+            .map_err(|err| calloop::Error::OtherError(Box::new(err)))?;
+
+        // Keyboard was destroyed
+        if removed {
+            return Ok(PostAction::Remove);
+        }
+
+        // Re-register the timer to start it again
+        if reregister {
+            return Ok(PostAction::Reregister);
+        }
+
+        let timer_pa = timer.process_events(readiness, token, |mut event, _| {
+            if self.disabled || key.is_none() {
+                // TODO How to pause the timer without dropping it?
+                // return TimeoutAction::ToDuration(Duration::from_millis(*delay_mut));
+                return TimeoutAction::Drop;
```

Comment:
bringing attention to this for @elinorbgr

Reply:
Dropping it and recreating a new one later is the intended usage on calloop, if that a problem?

Reply:
Ya, I guess, the issue was that that caused the PostAction to be PostAction::Remove iirc. But I think the logic that I have at line 213 for selecting a PostAction  works. I was just unsure that I had done it the best way.